### PR TITLE
Hash Join: variable radix bit counts

### LINKED ISSUE #1045: Find better _radix_bits
Comparing `_radix_bits = 9` to `_radix_bits = 0`:

```
+-----------+------------------+------+-----------------+------+---------+---------------------------------+
| Benchmark | prev. iter/s     | runs | new iter/s      | runs | change  | p-value (significant if <0.001) |
+-----------+------------------+------+-----------------+------+---------+---------------------------------+
| TPC-H 1   | 1.77056348324    | 213  | 1.75972640514   | 212  | -1%     |                          0.0054 |
| TPC-H 2   | 29.8127670288    | 3578 | 48.1226806641   | 5775 | +61%    |                          0.0000 |
| TPC-H 3   | 20.0760822296    | 2410 | 26.5160713196   | 3182 | +32%    |                          0.0000 |
| TPC-H 4   | 0.210900396109   | 26   | 0.212299197912  | 26   | +1%     |                          0.2118 |
| TPC-H 5   | 13.5522403717    | 1627 | 17.2646770477   | 2072 | +27%    |                          0.0000 |
| TPC-H 6   | 59.9964408875    | 7200 | 60.3719596863   | 7245 | +1%     |                          0.0000 |
| TPC-H 7   | 4.98639345169    | 599  | 5.51244926453   | 662  | +11%    |                          0.0000 |
| TPC-H 8   | 3.92327094078    | 471  | 2.94422745705   | 354  | -25%    |                          0.0000 |
| TPC-H 9   | 2.18278455734    | 262  | 2.23703861237   | 269  | +2%     |                          0.0000 |
| TPC-H 10  | 14.7751970291    | 1774 | 17.1352806091   | 2057 | +16%    |                          0.0000 |
| TPC-H 11  | 43.4202919006    | 5211 | 42.485622406    | 5099 | -2%     |                          0.0000 |
| TPC-H 12  | 21.1038837433    | 2533 | 30.5092163086   | 3662 | +45%    |                          0.0000 |
| TPC-H 13  | 12.131726265     | 1456 | 12.6561632156   | 1519 | +4%     |                          0.0000 |
| TPC-H 14  | 47.1568641663    | 5659 | 67.6981430054   | 8124 | +44%    |                          0.0000 |
| TPC-H 15  | 0.245661795139   | 30   | 51.6224479675   | 6195 | +20914% |                          0.0000 |
| TPC-H 16  | 9.42947769165    | 1132 | 10.8437166214   | 1302 | +15%    |                          0.0000 |
| TPC-H 17  | 1.95363533497    | 235  | 1.8807322979    | 226  | -4%     |                          0.0000 |
| TPC-H 18  | 0.0124519411474  | 2    | 1.78255414963   | 214  | +14215% |        (not enough runs) 0.0000 |
| TPC-H 19  | 2.05173134804    | 247  | 1.95750486851   | 235  | -5%     |                          0.0000 |
| TPC-H 20  | 0.00828054361045 | 1    | 0.0082748318091 | 1    | -0%     |           (not enough runs) nan |
| TPC-H 21  | 0.126209974289   | 16   | 0.127985700965  | 16   | +1%     |                          0.0722 |
| TPC-H 22  | 0.479876816273   | 58   | 0.480097740889  | 58   | +0%     |                          0.9111 |
| average   |                  |      |                 |      | +1607%  |                                 |
+-----------+------------------+------+-----------------+------+---------+---------------------------------+
```

Scale factor was `0.1`. For small tables, it is expected that a low number of buckets works better than `2^9=512`, but this...

And yes, all tests pass.

### PR DESCRIPTION
**Pull request by Bouncner**
New PR for the automated setting of radix bit counts for the hash join. fixes #1045

The reviews comments from PR #1049 should all be addressed.

TPC benchmark: `hyriseBenchmarkTPCH --scale 0.1 --time=720`
```
+-----------+---------------------+------+---------------------+------+---------+---------------------------------+
| Benchmark | prev. iter/s        | runs | new iter/s          | runs | change  | p-value (significant if <0.001) |
+-----------+---------------------+------+---------------------+------+---------+---------------------------------+
| TPC-H 1   | 2.72465515136719    | 1000 | 2.74556469917297    | 1000 | +1%     |                          0.0000 |
| TPC-H 2   | 21.2441997528076    | 1000 | 43.3155403137207    | 1000 | +104%   |     (run time too short) 0.0000 |
| TPC-H 3   | 19.9956378936768    | 1000 | 32.0532913208008    | 1000 | +60%    |     (run time too short) 0.0000 |
| TPC-H 4   | 0.310515403747559   | 224  | 0.311020076274872   | 224  | +0%     |                          0.4757 |
| TPC-H 5   | 15.5644903182983    | 1000 | 21.7481861114502    | 1000 | +40%    |     (run time too short) 0.0000 |
| TPC-H 6   | 50.2312278747559    | 1000 | 49.9696350097656    | 1000 | -1%     |     (run time too short) 0.0000 |
| TPC-H 7   | 4.71117353439331    | 1000 | 5.47914123535156    | 1000 | +16%    |                          0.0000 |
| TPC-H 8   | 3.26432061195374    | 1000 | 3.39146399497986    | 1000 | +4%     |                          0.0000 |
| TPC-H 9   | 1.90977990627289    | 1000 | 2.18742704391479    | 1000 | +15%    |                          0.0000 |
| TPC-H 10  | 15.5401010513306    | 1000 | 18.8029460906982    | 1000 | +21%    |     (run time too short) 0.0000 |
| TPC-H 11  | 43.7453155517578    | 1000 | 60.879035949707     | 1000 | +39%    |     (run time too short) 0.0000 |
| TPC-H 12  | 19.3144836425781    | 1000 | 27.0078945159912    | 1000 | +40%    |     (run time too short) 0.0000 |
| TPC-H 13  | 18.3297138214111    | 1000 | 19.1953983306885    | 1000 | +5%     |     (run time too short) 0.0000 |
| TPC-H 14  | 40.4100151062012    | 1000 | 67.9950332641602    | 1000 | +68%    |     (run time too short) 0.0000 |
| TPC-H 15  | 0.278847515583038   | 201  | 58.3825149536133    | 1000 | +20837% |     (run time too short) 0.0000 |
| TPC-H 16  | 6.90262794494629    | 1000 | 10.9350290298462    | 1000 | +58%    |                          0.0000 |
| TPC-H 17  | 1.37187445163727    | 988  | 1.36521983146667    | 983  | -0%     |                          0.0034 |
| TPC-H 18  | 0.0323572233319283  | 24   | 0.237914875149727   | 172  | +635%   |                          0.0000 |
| TPC-H 19  | 1.75965893268585    | 1000 | 1.85826230049133    | 1000 | +6%     |                          0.0000 |
| TPC-H 20  | 0.00561679294332862 | 5    | 0.00561118358746171 | 5    | -0%     |        (not enough runs) 0.5985 |
| TPC-H 21  | 0.185014992952347   | 134  | 0.185903683304787   | 134  | +0%     |                          0.0016 |
| TPC-H 22  | 0.307335644960403   | 222  | 0.302841901779175   | 219  | -1%     |                          0.0000 |
| average   |                     |      |                     |      | +998%   |                                 |
+-----------+---------------------+------+---------------------+------+---------+---------------------------------+
```

### CODE REVIEW

**Comment by Bouncner** [comment]
Addresses #1045 .

I'll do some further benchmarks of queries which seems to be faster with a static bit count (e.g., query 18).
My benchmarks ran on the rapa.

File: src/lib/operators/join_hash.cpp [hunk 1]
```diff
@@ -388,6 +391,13 @@ void probe(const RadixContainer<RightType>& radix_container,
       PosList pos_list_left_local;
       PosList pos_list_right_local;
 
+      // simple heuristic to estimate result size: half of the partition's row will match
+      // a more conservative pre-allocation would be the size of the left cluster
+      // TODO: info is in the radix_container ...
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: ?

File: src/lib/operators/join_hash.cpp [hunk 2]
```diff
@@ -617,8 +627,49 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+            be expected to fit into the L2 cache.
+            This should incorporate hardware knowledge, once available in Hyrise.
+            As of now, we assume a L2 cache size of 256 KB.
+            We estimate the size the following way:
+              - we assume each key appears once (that is an overestimation space-wise, but we
+              aim rather for a hash map that is slightly smaller than L2 than slightly larger)
+              - each entry in the hash map is a data structure holding the actual value
+              and the RowID
+          */
+          const auto build_relation_size = inputs_swapped ? _right->get_output()->row_count() : _left->get_output()->row_count();
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Please verify, but I think that `inputs_swapped` means that the inputs already *were* swapped, not that you have to do it again.

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (Bouncner) [reply src/lib/operators/join_hash.cpp]: Yeah, I wanted to start a "changes requested" review for that line, but that's not possible. I am on it.

Comment (Bouncner) [reply src/lib/operators/join_hash.cpp]: Addressed by removing TODO. Maybe later address the sizing when we find the current reserve to be too large or too small.

File: src/lib/operators/join_hash.cpp (hunk 2 above)
Comment (Bouncner) [reply src/lib/operators/join_hash.cpp]: Yep, right. Done.

**Pull request merged**